language/stdlib/modules/libra_system_2.move

```diff
@@ -0,0 +1,264 @@
+module LibraSystem2 {
+    use 0x0::LibraAccount;
+    use 0x0::ValidatorConfig2;
+    use 0x0::Vector;
+    use 0x0::Transaction;
+    use 0x0::LibraTimestamp;
+
+    struct ValidatorInfo {
+        addr: address,
+        consensus_voting_power: u64,
+        config: ValidatorConfig2::Config,
+        // To protect against DDoS, the force updates are allowed
+        // only if the time passed since the last force update is larger than a threshold (24hr).
+        last_force_update_time: u64,
+    }
+    
+    struct ValidatorSetChangeEvent {
+        new_validator_set: vector<ValidatorInfo>,
+    }
+
+    resource struct ValidatorSet {
+        validators: vector<ValidatorInfo>,
+        last_reconfiguration_time: u64,
+        change_events: LibraAccount::EventHandle<ValidatorSetChangeEvent>,
+    }
+    
+    // ValidatorInfo public accessors
+    public fun get_validator_address(v: &ValidatorInfo): &address { &v.addr }
+    public fun get_validator_config(v: &ValidatorInfo): &ValidatorConfig2::Config { &v.config }
+    public fun get_validator_voting_power(v: &ValidatorInfo): &u64 { &v.consensus_voting_power }
+    public fun get_last_force_update_time(v: &ValidatorInfo): &u64 { &v.last_force_update_time }
+
+    // This can only be invoked by the ValidatorSet address to instantiate the resource under that address.
+    // It can only be called a single time. Currently, it is invoked in the genesis transaction.
+    public fun initialize_validator_set() {
+        Transaction::assert(Transaction::sender() == 0x1D8, 1);
+        move_to_sender<ValidatorSet>(ValidatorSet {
+            validators: Vector::empty(),
+            last_reconfiguration_time: 0,
+            change_events: LibraAccount::new_event_handle<ValidatorSetChangeEvent>(),
+        });
+    }
+    
+    // Adds a new validator, only callable by the LibraAssociation address
+    // TODO(valerini): allow the association to add multiple validators in a single block
+    public fun add_validator(account_address: &address) acquires ValidatorSet {
+        Transaction::assert(Transaction::sender() == 0xA550C18, 1);
+        // A prospective validator must have a validator config resource
+        Transaction::assert(ValidatorConfig2::has(*account_address), 17);
+        
+        let validator_set_ref = borrow_global_mut<ValidatorSet>(0x1D8);
+        // Ensure that this address is not already a validator
+        Transaction::assert(!is_validator_(account_address, &validator_set_ref.validators), 18);
+
+        let config = ValidatorConfig2::get_config(*account_address);
+        let current_block_time = LibraTimestamp::now_microseconds();
+        Vector::push_back(&mut validator_set_ref.validators, ValidatorInfo {
+            addr: *account_address,
+            config: *&config, // copy the config over to ValidatorSet
+            consensus_voting_power: 1,
+            last_force_update_time: current_block_time,
+        });
+        
+        emit_reconfiguration_event();
```

Comment:
do we want to couple these events together? same with add / remove... I'd love for us to have an explicit ask for a reconfiguration event as part of a script ... like try_emit_reconfiguration_event... if we don't want it, we don't call it and we wait for either an eventual reconfiguration or someone else....

Reply:
Yes, it would be very nice to couple these events together!! And I agree that reconfiguration should be preferably explicitly triggered.

Perhaps using the idea in Zekun's [PR3308](https://github.com/libra/libra/pull/3308) (ref. @zekun000), we can keep the ValidatorSet in LibraSystem and use it as a temporal buffer to add, remove, update validators, and copy this set over to LibraConfig upon an explicit call and that is when the reconfiguration event will get emitted?